<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{R/simcross User Guide}
-->

# R/simcross User Guide

[R/simcross](https://github.com/kbroman/simcross) is an
[R](http://www.r-project.org) package to simulate genotypes from an
experimental cross. The aim is flexibility rather than speed.

Meiosis is simulated following the Stahl model (Copenhaver et
al. 2002) with the interference parameter being an integer. In this
model, chiasmata locations are the superposition of two processes: a
proportion `p` come from a process exhibiting no interference, and the
remainder (proportion `1-p`) come from a process following the
chi-square model (Foss et al. 1993; Zhao et al. 1995). This, with
`p=0`, the model reduces to the chi-square model. The case `m=0` (or
`p=1`) corresponds to the no interference model.  Broman et al. (2002)
estimated `m=10` for the level of interference in the mouse (assuming
`p=0`).

## Basic functions

There are two basic functions for simulating a cross:
`create_parent` and `cross`.

`create_parent` parent is for generating a parent object, either an
inbred individual or the F<sub>1</sub> offspring of two inbred
individuals. It takes two arguments: the length of the chromosome in
cM, and the allele or pair of alleles.

Here's how to generate two inbred individuals, one from each of strain
1 and strain 2, and an F<sub>1</sub> individual.

```{r set.seed, echo=FALSE, results="hide"}
set.seed(80605735)
```

```{r create_parent}
library(simcross)
p1 <- create_parent(100, 1)
p2 <- create_parent(100, 2)
f1 <- create_parent(100, 1:2)
```

The `cross` function is used to generate one progeny from the cross
between two individuals.  The input is a pair of individuals (e.g., as
produced by `create_parent`), the interference parameter `m` (default
`m=10`), the parameter `p` for the Stahl model (default `p=0`), `xchr`
to indicate whether the chromosome being simulated is the X chromosome
(`FALSE`, the default, to simulate an autosome), and `male` to
indicate whether the offspring is to be male (which only matters if
`xchr=TRUE`).

Here's an example, to generate an F<sub>2</sub> individual.

```{r sim_f2}
f2 <- cross(f1, f1)
print(f2)
```

The output is list with two components: the maternal and paternal
chromosomes. Each chromsome is a data frame with the allele in a set
of intervals, and the locations of the right endpoints of the
intervals.

In the example above, the maternal chromosome is a non-recombinant `1`
chromosome. The paternal chromosome has two crossovers, with the
the `2` allele up to `r round(f2$pat$locations[1], 1)` cM, the `1` allele
in the interval `r round(f2$pat$locations[1], 1)` &ndash;
`r round(f2$pat$locations[2], 1)`, and the `2` allele
for the remainder of the chromosome.

There's also a function `meiosis` for simulating a meiotic product
from a single individual; the `cross` function calls `meiosis` twice
(once for each parent) and then combines the results into a single
individual.

## Generating pedigrees

The functions `sim_ril_pedigree`, `sim_ail_pedigree`, and
`sim_do_pedigree`.

## Simulating from a pedigree

The function `sim_from_pedigree`.

## Generating marker data

The function `convert2geno` and `get_geno`.




## References

Broman KW, Rowe LB, Churchill GA, Paigen K (2002)
[Crossover interference in the mouse](http://www.genetics.org/content/160/3/1123.short).
Genetics 160:1123-1131

Copenhaver, G. P., Housworth, E. A. and Stahl, F. W. (2002)
[Crossover interference in arabidopsis](http://www.genetics.org/content/160/4/1631.short).
Genetics 160:1631-1639.

Foss, E., Lande, R., Stahl, F. W. and Steinberg, C. M. (1993)
[Chiasma interference as a function of genetic distance](http://www.genetics.org/content/133/3/681.short).
Genetics 133:681-691.

Zhao H, Speed TP, McPeek MS (1995)
[Statistical analysis of crossover interference using the chi-square model](http://www.genetics.org/content/139/2/1045.short).
Genetics 139:1045-1056

<!-- the following to make it look nicer -->
<link href="http://kbroman.org/qtlcharts/assets/vignettes/vignette.css" rel="stylesheet"></link>
