# convert2geno:
#
#' Convert continuous allele information into marker genotypes
#'
#' Convert the continuous crossover location information produced by
#' sim_from_pedigree to marker genotypes
#'
#' @param xodat The sort of detailed genotype/crossover data generated by
#' \code{\link{sim_from_pedigree}}
#' @param map vector of marker locations
#' @param founder_geno Optional matrix (size \code{n_founders} x
#' \code{length(map)}) of founder genotypes coded as 1/2. If provided,
#' results are 1/2/3 genotypes
#'
#' @return If \code{founder_geno} is provided or there are just two
#' founders, the result is a numeric matrix of genotypes, individuals
#' x markers, with genotypes 1/2/3 codes for 11/12/22 genotypes.
#'
#' If \code{founder_geno} is not provided and there are more than two
#' founders, the result is a 3-dimensional array, individuals x
#' markers x alleles, with the third dimensional corresponding to the
#' maternal and paternal allele.
#'
#' @export
#' @keywords utilities
#' @seealso \code{\link{get_geno}}, \code{\link{sim_from_pedigree}}
#'
#' @examples
#' # simulate AIL pedigree
#' tab <- sim_ail_pedigree(12, 30)
#' # simulate data from that pedigree
#' dat <- sim_from_pedigree(tab)
#' # marker map (could also use sim.map in R/qtl)
#' map <- seq(0, 100, by=5)
#' names(map) <- paste0("marker", seq(along=map))
#' # convert data to marker genotypes
#' geno <- convert2geno(dat, map)
#'
#' # simulate DO pedigree
#' tab <- sim_do_pedigree(8)
#' # simulate data from that pedigree
#' dat <- sim_from_pedigree(tab)
#' # simulate founder snp alleles
#' fg <- matrix(sample(1:2, 8*length(map), repl=TRUE), nrow=8)
#' # for DO, need female & male founders (to deal with X chr)
#' fg <- rbind(fg, fg)
#' # convert dat to SNP genotypes
#' geno <- convert2geno(dat, map, fg)
#' # if fg not provided, result is a 3d array
#' genoarray <- convert2geno(dat, map)
convert2geno <-
function(xodat, map, founder_geno)
{
  if(map[1] != 0) map <- map - map[1]
  if(max(map) > max(xodat[[1]]$mat$locations))
    warning("maximum simulated position is less than the length of the map.")

  # maximum founder genotype
  max_geno <- max(sapply(xodat, function(a) max(a$mat$alleles, a$pat$alleles)))

  if(missing(founder_geno) || is.null(founder_geno))
      founder_geno <- matrix(nrow=0, ncol=0)
  else {
      if(nrow(founder_geno) < max_geno)
          stop("founder_geno should have at least ", max_geno,
               " rows but has ", nrow(founder_geno))
      if(ncol(founder_geno) != length(map))
          stop("founder_geno should have ", length(map),
               "columns but has ", ncol(founder_geno))
      if(any(is.na(founder_geno)))
          stop("founder genotypes cannot be missing")
      if(any(founder_geno != 1 & founder_geno != 2))
          stop("founder_geno must contain only 1's and 2's")
  }

  if(length(founder_geno) > 0 || max_geno <= 2) {
      output <- t(.Call('simcross_fromR_convert2geno', PACKAGE = 'simcross',
                        xodat, map, founder_geno))

      dimnames(output) <- list(names(xodat), names(map))
  }
  else {
      output <- .Call('simcross_fromR_convert2genoarray', PACKAGE = 'simcross',
                      xodat, map)
      # output was markers x ind x alleles; switch first two dimensions
      output <- aperm(output, c(2, 1, 3))

      dimnames(output) <- list(names(xodat), names(map), c("mat", "pat"))
  }

  output
}
